server/computer_use/handlers/opencua.py

```diff
@@ -0,0 +1,550 @@
+            else:
+                data = '{}'
+
+            if data:
+                try:
+                    data = json.loads(data)
+                except json.JSONDecodeError:
+                    logger.warning(f'Invalid JSON in data: {data}')
+
+            print('OpenCua terminate status', repr(status))
+            print('OpenCua terminate data', repr(data))
+
+            if status == 'success':
+                return {
+                    'id': 'toolu_opencua_terminate',
+                    'type': 'tool_use',
+                    'name': 'extraction',
+                    'input': {'data': {'status': status, 'data': data}},
+                }
+            else:
+                # handle if data is not a dict
+                if not isinstance(data, dict) or 'reasoning' not in data:
+                    data = {'reasoning': data}
+
+                return {
+                    'id': 'toolu_opencua_terminate',
+                    'type': 'tool_use',
+                    'name': 'ui_not_as_expected',
+                    'input': {
+                        'data': {'status': status, 'reasoning': data['reasoning']}
+                    },
+                }
+
+        raise ValueError(f'Unknown command: {command}')
+
+    def convert_from_provider_response(
+        self, response: str
+    ) -> tuple[list[BetaContentBlockParam], str]:
+        """Convert OpenCua response to provider-specific response."""
+
+        print('OpenCua response', repr(response))
+
+        task = self._parse_task(response)
+
+        print('OpenCua task', repr(task))
+
+        messages: list[BetaContentBlockParam] = []
+        stop_reason = 'end_turn'
+
+        # Only keep the step and action (alligns with samples from the openCua paper);
+        # TODO: Having the though displayed in the UI is nice, but hurts model performance
+        # -> include and filter out in convert_to_provider_messages
+
+        # construct text message from task
+        text_message = ''
+        if task['step']:
+            text_message += f'# Step: {task["step"]}\n'
+        if task['action']:
+            text_message += f'## Action: {task["action"]}\n'
+
+        if text_message:
+            messages.append(
+                cast(BetaTextBlockParam, {'type': 'text', 'text': text_message})
+            )
+
+        if task['code']:
+            tool_use = self._convert_pyautogui_code_to_tool_use(task['code'])
+            messages.append(tool_use)
+
+            # End the turn once extraction or ui_not_as_expected is called
+            if tool_use['id'] != 'toolu_opencua_terminate':
+                stop_reason = 'tool_use'
+
+        # thought can be dropped, as it's is more or less just reasoning output of the model itself
+        # The paper does not include it in the message history, but for the user it might be of interest
+
+        print('OpenCua messages', repr(messages), 'stop_reason', stop_reason)
+
+        return messages, stop_reason
```

Comment:
might make sense to have a module with multiple files for each handler, this way we could extract individual features and a lot of lines (e.g. system prompt) to dedicated files and improve readability of the code